### PR TITLE
Prevent doubling stake

### LINKED ISSUE #185: Double stake amount when staking 0 amount to unlocked stake
issue happnes in following situation:
- `LockupPeriod`: 10
- holder A stakes 1000 AMO to validator A -> locked at height 10
- at height 20, stake of 1000 AMO gets unlocked
- holder A stakes 0 AMO to validator A - locked at height 21
- total stake amount of holder A gets doubled, resulting in 2000 AMO

problem is caused in following code:
```go
// check balance
	balance := store.GetBalance(t.GetSender(), false)
	if balance.LessThan(&txParam.Amount) {
		return code.TxCodeNotEnoughBalance, "not enough balance", nil
	}

	balance.Sub(&txParam.Amount)
	stake := store.GetStake(t.GetSender(), false)
	if stake == nil {
		var k ed25519.PubKeyEd25519
		copy(k[:], txParam.Validator)
		stake = &types.Stake{
			Amount:    txParam.Amount,
			Validator: k,
		}
	} else if bytes.Equal(stake.Validator[:], txParam.Validator[:]) {
		stake.Amount.Add(&txParam.Amount)
	} else {
		return code.TxCodePermissionDenied, "validator key mismatch", nil
	}
	err = store.SetLockedStake(t.GetSender(), stake, int64(ConfigLockupPeriod))
```
- bypass error check processes
- too loose balance check process: balance > 0 == true (always if balance is bigger than 0)
- `balance.Sub(&txParam.Amount)`: balace - 0 -> balance
- `stake := store.GetStake(t.GetSender(), false)`: only total stake amount
- `stake.Amount.Add(&txParam.Amount)`:  stake(1000AMO) + 0 -> another stake(1000AMO)
- `err = store.SetLockedStake(t.GetSender(), stake, int64(ConfigLockupPeriod))`
  - the stake(1000AMO) is haneled as another stake and locked at given height
  - unlocked stake(1000AMO) + locked stake(1000AMO) -> 2000AMO

This kind of 'Double stake with 0 amount' doesn't happen in `transfer`, `delegate` txs, as in `transfer` tx, the balance of `sender` and `receiver` is checked and compared and in `delegate` tx, there is no unlocked or locked delegate.

### PR DESCRIPTION
**Pull request by h0n9**
- set conditions for checking 0 amount while executing txs
- handle stake tx only as a newly generated locked stake
- add test codes for stake and overall txs

resolves https://github.com/amolabs/amoabci/issues/185

### CODE REVIEW

**Comment by h0n9** [comment]
update for docs in https://github.com/amolabs/docs/commit/d9b0e640bc0654f01ecc47273a431db32999612b

File: amo/code/code.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ const (
 	TxCodeBadParam
 	TxCodeTooOldTx
 	TxCodeAlreadyProcessedTx
+	TxCodeUnavailableAmount
```
Comment (Lbird) [inline amo/code/code.go]: InvalidAmount, and change rest of the codes.

File: amo/tx/delegate.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ func (t *TxDelegate) Execute(store *store.Store) (uint32, string, []tm.KVPair) {
 		return code.TxCodeBadParam, err.Error(), nil
 	}
 
+	if !txParam.Amount.GreaterThan(zero) {
+		return code.TxCodeUnavailableAmount, "unavailable amount", nil
```
Comment (Lbird) [inline amo/tx/delegate.go]: invalid amount

File: amo/tx/stake.go [hunk 3]
```diff
@@ -53,25 +52,24 @@ func (t *TxStake) Execute(store *store.Store) (uint32, string, []tm.KVPair) {
...
 	balance.Sub(&txParam.Amount)
-	stake := store.GetStake(t.GetSender(), false)
-	if stake == nil {
-		var k ed25519.PubKeyEd25519
-		copy(k[:], txParam.Validator)
-		stake = &types.Stake{
-			Amount:    txParam.Amount,
-			Validator: k,
-		}
-	} else if bytes.Equal(stake.Validator[:], txParam.Validator[:]) {
```
Comment (Lbird) [inline amo/tx/stake.go:83]: we should not add something, but we should check validator match and return BadValidator or ValidatorMismatch when it does not match. need to check somewhere. check these:
- stake owner mismatch (permission denied)
- validator key mismatch (validator mismatch)

File: amo/tx/tx.go [hunk 4]
```diff
@@ -19,12 +19,13 @@ const (
 )
 
 var (
-	c = elliptic.P256()
+	// config values from the app
+	ConfigLockupPeriod = defaultLockupPeriod
```
Comment (Lbird) [inline amo/tx/tx.go]: This will raise conflict when merging. merge from the recent master branch to avoid conflict.

**Review by Lbird — CHANGES_REQUESTED** [review]

**Review by Lbird — APPROVED** [review]

**Pull request merged**